### PR TITLE
V3 baseline

### LINKED ISSUE #539: Allow reusing of files in page layouts
**Version: v1.15.2**

**Problem**

Currently page layouts would need to duplicate files if the exact same files are used as layouts are treated as stand-alone folders. 

**Potential solution**

> Introduce a new file e.g., settings.md into layouts that can contain this kind of settings?

To extend your idea, currently the layout files for footer and other items are separated. It would be neater if all the different files (footer, siteNav, css) are all stored in their respective folders and have a simple `settings.md` file to "store" the file names required. Basically abstracting layouts to just a front matter file.

E.G, content within `settings.md`:
```
css: colorTheme1.css, blackTheme2.css
footer: simpleFooter.md
pageNav: default
pageNavHeader: "This is a list of stuff"
siteNav: mySiteNav.md
script: myScript1.js, myScript2.js
```

Then all the different file types are just stored in their respective folders within `_/markbind`

Original comment by @Chng-Zhi-Xuan in https://github.com/MarkBind/markbind/pull/465#issuecomment-451625452_

### LINKED ISSUE #810: Refactor the Parser/Site/Page trinity into more comprehensible components
While doing so, also explore the possibility of parsing .md files only once (see #762 for more context)

In cases where complexity cannot be avoided, make sure to add comments to the relevant components/functions and document any unintuitive behaviour. (Related to #602)

### LINKED ISSUE #972: Footnotes inside panels don't work
Example (see the footnote inside the _tutorials_ panel) https://nus-cs2113-ay1920s2.github.io/website/admin/weeklySchedule.html

image

### LINKED ISSUE #973: Support search bar in siteNav
v2.9.0
Current: the search bar is in the topNav
Problem: user needs to scroll to top before searching
Suggestion: provide a way to insert a search bar into the siteNav (which has a sticky behavior compared to topNav)

### LINKED ISSUE #1188: Add more hooks for plugins
**Is your request related to a problem?**

Work done here for a rough idea: #968 

Plugins modify the content currently only through 2 hooks, `pre/postRender`.
This is very expensive however, as it involves parsing each page an additional time for each plugin there is.

**Describe the solution you'd like**

Since element processing in `componentParser/Preprocessor` is now split cleanly on a per-element basis, and these elements are presented in a familiar format that is from `htmlParser2`, we could provide hooks in these places to let plugins tap directly into the processes(s) here.

Cons:
- It's still a less clean solution than strictly abiding by passing the string `content`, but
  - Reparsing `content` for every plugin there is isn't really scalable in achieving #984 
- `componentPreprocessor` has some amount of cyclic dependencies with `Parser` still that shouldn't need to be exposed to plugins ideally. We could find a solution to this first, or only do it for `componentParser`

**Describe alternatives you've considered**

na

**Additional context**

na

### LINKED ISSUE #1240: layouts: don't include files that are missing
v2.14.1

Current: Even if a layout is missing a standard file e.g., styles.css, a link to the file is inserted into every page, resulting in the browser reporting 'missing file'. See screenshot (results of a tool checking for broken links)

image

Suggestion: don't insert a link to such files if the file is missing.

### LINKED ISSUE #1308: Clarify the requirements of layouts
Currently, layouts (header, footer, page (expressive) layouts, navigation) cannot:
- [ ] use include features (`<panel src="...">` & `<include>`) (with the exception of expressive layouts)
- [ ] use `<variable>` / `<import>` correctly (variables are not erased, imports do not work with live reload)
- [ ] use plugin elements correctly (as `preRender` is not executed for layouts)
- ❔ 

Is this by design? 

If not, a very large refactor may be due to fix the processing order

### PR DESCRIPTION
**Pull request by ang-zeyu**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [x] Others, please explain:

The architecture involves too much redundant processing; And the order of these processing stages is far too coupled.

This PR revolves around following this architecture: nunjucks -> markdown -> html, and also overhauls the layouts system in favour of one-file expressive layouts per #1308 at the same time (only because it makes the rewrite simpler =P)

Some plugin interface tweaks inline / required by the architecture simplification are also done

Resolves #1308 
Resolves #1240 
Resolves #539 
Resolves #1188 
Resolves #973 
Resolves #810 
Fixes #972 

Related:
#896 

**Overview of changes:**
- oop-fied codebase (new manager classes: `LayoutManager`, `ExternalManager`, `PluginManager` new object classes `Layout`, `External`, `Plugin`)
- simpler architecture that always follows: nunjucks -> markdown -> html
   - `nodePreprocessor` and `nodeProcessor` is combined into the html stage
- plugin system tweaks
- major performance improvements ~3x
- updated layouts docs

**Anything you'd like to highlight / discuss:**

Todos:
- all done

**Testing instructions:**
manually tested layouts system

**Proposed commit message: (wrap lines at 72 characters)**
Rebase - commits need to be merged as one but are quite distinct

---

**Checklist:** :ballot_box_with_check:

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [ ] No blatantly unrelated changes 🤔 bigger than I'd like but its far too coupled to deliver separately
- [x] Pinged someone for a review!

### CODE REVIEW

**Comment by ang-zeyu** [comment]
@damithc @acjh  any thoughts on the layouts changes?

https://deploy-preview-1403--markbind-master.netlify.app/userguide/tweakingthepagestructure

So far I've noticed:
- additional work for fixed headers is now needed - the `fixed-header-padding` class must be added by the user to the appropriate elements
- a heavy shift of layout related styles to userland - e.g. `#content-wrapper`, `#site-nav`, .... This now becomes the user's responsibility to maintain. (https://github.com/MarkBind/markbind/pull/1403/files#diff-309b964023ca899c9505205f36d3f4d5b36a6487e5c9b2e242204ee06bbc9ce9)

Its something we can reduce the work needed in template sites though, and may be warranted by the flexibility provided. (many interesting layouts will be made possible by this)

Not sure if there are more alternatives to alleviate this though

**Comment by damithc** [comment]
> * additional work for fixed headers is now needed - the `fixed-header-padding` class must be added by the user to the appropriate elements

In every page? Or just in the layout pages?


> * a heavy shift of layout related styles to userland - e.g. `#content-wrapper`, `#site-nav`, .... This now becomes the user's responsibility to maintain. (https://github.com/MarkBind/markbind/pull/1403/files#diff-309b964023ca899c9505205f36d3f4d5b36a6487e5c9b2e242204ee06bbc9ce9)

I noticed `<span id="title" class="d-none">{{ title }}</span>` being added to many pages. Is it quick of your own site or something authors have to do in every page in every site?

File: docs/userGuide/usingComponents.md [hunk 1]
```diff
@@ -1,9 +1,10 @@
-<variable name="title" id="title">Using Components</variable>
```
Comment (ang-zeyu) [inline docs/userGuide/usingComponents.md:1]: > In every page? Or just in the layout pages?

Just layouts : )

Still something extra to take note of for the author though

> > * a heavy shift of layout related styles to userland - e.g. `#content-wrapper`, `#site-nav`, .... This now becomes the user's responsibility to maintain. (https://github.com/MarkBind/markbind/pull/1403/files#diff-309b964023ca899c9505205f36d3f4d5b36a6487e5c9b2e242204ee06bbc9ce9)
> 
> I noticed `<span id="title" class="d-none">{{ title }}</span>` being added to many pages. Is it quick of your own site or something authors have to do in every page in every site?

This one would be belong to the page `<variable>` / `<import>` removal pr (to be split), there's no way to set the id for a nunjucks tag so had to create another `<span>`. The title `id` is used for the userguide macro that creates the left / right navigation buttons at the bottom

Comment (damithc) [reply docs/userGuide/usingComponents.md:1]: Got it.

**Comment by ang-zeyu** [comment]
Any thoughts on the docs / feature set? @damithc 

https://github.com/MarkBind/markbind/pull/1403/commits/44d81670e7fba7edbccb4c7c81858d3e15351616

I'll give the code a few more looks here and there this week, then maybe merge it in this weekend otherwise.

**Comment by damithc** [comment]
> Any thoughts on the docs / feature set? @damithc

Looks fine for me :+1: 
Any idea how much change this will require in CS2103 website?

**Comment by ang-zeyu** [comment]
> Looks fine for me 👍
> Any idea how much change this will require in CS2103 website?

Hmm from what I can see there aren't a lot of layouts, so most of the work should be from converting page `<variable>` / `<import>`s to nunjucks `{% set / import %}`. Would depend on how many of those you have =X

For layouts you can re-run `markbind init` and copy over the default layout as a baseline, then 'simply `<include>`' them, and ctrl-R the layout names

Also the old attribute names which have been deprecated for a while now (`title`, `text`, ...) may be removed in a later PR (or we can postpone maybe, I don't think its affecting maintenance too much)

**Comment by ang-zeyu** [comment]
Merging this in a little earlier in case anyone starts looking at the backend.

\*fingers crossed\*, but we can find (hopefully) and fix any issues before release

**Pull request merged**